src/Algebra/Properties/CommutativeMonoid.agda

```diff
@@ -0,0 +1,168 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some derivable properties
+------------------------------------------------------------------------
+
+open import Algebra
+
+module Algebra.Properties.CommutativeMonoid {g₁ g₂} (M : CommutativeMonoid g₁ g₂) where
+
+open CommutativeMonoid M renaming (ε to 0#; _∙_ to _+_; ∙-cong to +-cong; identity to +-identity; assoc to +-assoc; comm to +-comm)
+open import Algebra.Operations.CommutativeMonoid M
+import Algebra.FunctionProperties as Props; open Props _≈_
+import Relation.Binary.EqReasoning as EqR; open EqR setoid
+import Relation.Binary as B
+open import Function
+open import Function.Equality using (_⟨$⟩_)
+open import Data.Product
+import Data.Bool as Bool
+open import Data.Nat using (ℕ; zero; suc)
+open import Data.Fin using (Fin; punchIn; zero; suc)
+open import Data.List as List using ([]; _∷_)
+import Data.Fin.Properties as FP
+open import Data.Fin.Permutation as Perm using (Permutation; Permutation′; _⟨$⟩ˡ_; _⟨$⟩ʳ_)
+open import Data.Fin.Permutation.Components as PermC
+open import Data.Table as Table
+open import Data.Table.Relation.Equality as TE using (_≗_)
+open import Data.Unit using (tt)
+open import Data.Empty using (⊥-elim)
+import Data.Table.Properties as TP
+open import Relation.Binary.PropositionalEquality as P using (_≡_)
+open import Relation.Nullary using (yes; no)
+open import Relation.Nullary.Decidable using (⌊_⌋)
+
+module _ {n} where
+  open B.Setoid (TE.setoid setoid n) public
+    using ()
+    renaming (_≈_ to _≋_)
+
+-- When summing over a function from a finite set, we can pull out any value and move it to the front.
+sumₜ-punchIn : ∀ {n} t (i : Fin (suc n)) → sumₜ t ≈ lookup t i + sumₜ (rearrange (punchIn i) t)
+sumₜ-punchIn f zero = refl
+sumₜ-punchIn {zero} t (suc ())
+sumₜ-punchIn {suc n} t (suc i) =
+  begin
+    head t + sumₜ (tail t)                                                ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
+    head t + (lookup t (suc i) + sumₜ (rearrange (punchIn i) (tail t)))   ≈⟨ sym (+-assoc _ _ _) ⟩
+    (head t + lookup t (suc i)) + sumₜ (rearrange (punchIn i) (tail t))   ≈⟨ +-cong (+-comm _ _) refl ⟩
+    (lookup t (suc i) + head t) + sumₜ (rearrange (punchIn i) (tail t))   ≈⟨ +-assoc _ _ _ ⟩
+    lookup t (suc i) + (head t + sumₜ (rearrange (punchIn i) (tail t)))   ∎
+
```

Comment:
IMO this whole proof would become a lot clearer if you started with a set of
let-bindings:

```agda
let x = head t
    y = lookup t (suc i)
    z = sumₜ (rearrange (punchIn i) (tail t))
in begin
￼    x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
￼    x + (y + z)        ≈⟨ sym (+-assoc x y z) ⟩
￼    (x + y) + z        ≈⟨ +-cong (+-comm x y) refl ⟩
￼    (y + x) + z        ≈⟨ +-assoc y x z ⟩
￼    x + (y + z)   ∎

Reply:
Good idea, done.